redirect-handler.js

```diff
@@ -68,18 +78,36 @@ module.exports = function buildRedirectHandler(filename, code = 301) {
   groupMatcher = new RegExp(`^(${escaped.join('|')})`);
 
   return (req, res, next) => {
+    let target;
+
     const url = ensureTrailingSlashOnly(req.path);
     if (url in singleRedirect) {
-      return res.redirect(code, singleRedirect[url]);
-    }
-
-    const m = groupMatcher.exec(url);
-    if (m && m[1] in groupRedirect) {
+      target = new URL(singleRedirect[url], baseUrl);
+    } else {
+      const m = groupMatcher.exec(url);
+      if (!(m && m[1] in groupRedirect)) {
+        return next();
+      }
       const base = groupRedirect[m[1]];
       const rest = url.slice(m[1].length);
-      return res.redirect(code, base + rest);
+      target = new URL(base + rest, baseUrl);
+    }
+
+    // Merge the original request's params into the target. We don't use
+    // Express' req.query here as it expands the query too much, e.g.,
+    // 'foo[bar]=123' ends up like "{foo: {bar: 123}}".
+    const requestParams = new URLSearchParams(req.url.substr(req.path.length));
+    for (const [key, value] of requestParams) {
+      target.searchParams.append(key, value);
     }
 
-    next();
+    // If the result URL starts with "https://web.dev/" (with trailing slash),
+    // then strip it and treat as local. This allows redirects to work in dev,
+    // where the domain is localhost or a staging URL.
+    let s = target.toString();
```

Comment:
Instead of doing this and hard coding the host on line 21, what if you did something like:

```
var url = require('url');

function baseUrl(req) {
  return url.format({
    protocol: req.protocol,
    host: req.get('host'),
  });
}
```

I think that should give you either localhost:8080 or web.dev.

Reply:
although I don't think that includes the port... so might need to grab that from app settings or something.

Reply:
I guess I didn't trust Express to give us the host properly. So the flip side is we don't do this at all, because I'm literally just trying to save a few bytes (rather than _always_ redirecting to https://web.dev/—I'm worrying too much).

Reply:
> I guess I didn't trust Express to give us the host properly.

Why's that? Not trying to be contrarian; genuinely curious.

> So the flip side is we don't do this at all, because I'm literally just trying to save a few bytes (rather than always redirecting to https://web.dev/—I'm worrying too much).

I could see it being useful to do. I've definitely ended up in situations where I thought I was developing locally and clicked on a link to went to the live site and didn't realize it and wondered why my changes weren't taking affect.

Reply:
I'm not worried about having a hard-coded "web.dev" string in the server, as we do it in a few places.